lib/_emerge/EbuildFetcher.py

```diff
@@ -273,17 +281,43 @@ def _target(settings, manifest, uri_map, fetchonly):
         if _want_userfetch(settings):
             _drop_privs_userfetch(settings)
 
-        rval = 1
         allow_missing = manifest.allow_missing or "digest" in settings.features
-        if fetch(
-            uri_map,
-            settings,
-            fetchonly=fetchonly,
-            digests=copy.deepcopy(manifest.getTypeDigests("DIST")),
-            allow_missing_digests=allow_missing,
-        ):
-            rval = os.EX_OK
-        return rval
+
+        async def main():
+            loop = asyncio.get_event_loop()
+            task = asyncio.ensure_future(
+                async_fetch(
+                    uri_map,
+                    settings,
+                    fetchonly=fetchonly,
+                    digests=copy.deepcopy(manifest.getTypeDigests("DIST")),
+                    allow_missing_digests=allow_missing,
+                )
+            )
+
+            def sigterm_handler(signum, _frame):
+                loop.call_soon_threadsafe(task.cancel)
+                signal.signal(signal.SIGTERM, signal.SIG_IGN)
+
+            signal.signal(signal.SIGTERM, sigterm_handler)
+            try:
+                await task
+            except asyncio.CancelledError:
+                # If asyncio.CancelledError arrives too soon after fork/spawn
+                # then handers will not have an opportunity to termintate
+                # the corresponding process, so clean up after this race.
+                for proc in multiprocessing.active_children():
```

Comment:
Why not loop multiple times over `active_children()`. Especially first sending SIGTERM to all children, and *then* joining them, gives the childs and the OS time to actually terminate the process correctly.

First time calls terminate, second time calls `join()`, where the first invocation of `join` has set time small timeout value, like 250ms. All others use 0 as explicit timeout value. And with the third and last iteration over `active_children()`, we kill all stragglers.

So this single loop becomes:

```python
for proc in multiprocessing.active_children():
    proc.terminate()

timeout = 0.25
for proc in multiprocessing.active_children():
     proc.join(timeout)
     timeout = 0

for proc in multiprocessing.active_children():
    proc.kill()
```

Reply:
Your version looks great except I'd add `proc.join()` after kill in order to ensure that the pid is reaped ASAP for reasons mentioned in https://github.com/gentoo/portage/pull/1361/files#r1688755132.

Reply:
Updated, thanks!

Reply:
Thanks, the added code comments make then intention now clear.